internals/overlord/logstate/gatherer.go

```diff
@@ -169,43 +175,48 @@ func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.R
 // pullers on entryCh, and writes them to the client. It also flushes the
 // client periodically, and exits when the gatherer's tomb is killed.
 func (g *logGatherer) loop() error {
-	timer := newTimer()
-	defer timer.Stop()
+	defer g.timer.Stop()
 
 mainLoop:
 	for {
 		select {
 		case <-g.tomb.Dying():
 			break mainLoop
 
-		case <-timer.Expired():
+		case <-g.timer.Expired():
 			// Mark timer as unset
-			timer.Stop()
-			err := g.client.Flush(g.clientCtx)
-			if err != nil {
-				logger.Noticef("Cannot flush logs to target %q: %v", g.targetName, err)
-			}
+			g.timer.Stop()
+			g.handleClientErr(g.client.Flush(g.clientCtx))
 
 		case entry := <-g.entryCh:
-			err := g.client.Write(g.clientCtx, entry)
-			if err != nil {
-				logger.Noticef("Cannot write logs to target %q: %v", g.targetName, err)
-			}
-			timer.EnsureSet(g.bufferTimeout)
+			g.handleClientErr(g.client.Write(g.clientCtx, entry))
+			g.timer.EnsureSet(g.bufferTimeout)
```

Comment:
A bit unclear why we are trying to pick up retry errors here?  The logic we implemented is Write, Write, Write, Timeout, Flush, so in theory it's that last step that should flush the buffer to the target. This makes it look like both Write and Flush are doing similar jobs.

Reply:
`Write` may call `Flush` when the buffer is full